lasagne/layers/autoenc.py

```diff
@@ -0,0 +1,262 @@
+import theano
+
+from .. import nonlinearities
+from .. import init
+
+from .base import MergeLayer
+from .helper import get_all_layers
+from .input import InputLayer
+from .special import NonlinearityLayer, BiasLayer
+from .noise import DropoutLayer, GaussianNoiseLayer
+
+
+__all__ = [
+    "InverseLayer",
+    "build_autoencoder",
+    "unfold_bias_and_nonlinearity_layers",
+]
+
+
+class InverseLayer(MergeLayer):
+    """
+    The :class:`InverseLayer` class performs inverse operations
+    for a single layer of a neural network by applying the
+    partial derivative of the layer to be inverted with
+    respect to its input: transposed layer
+    for a :class:`DenseLayer`, deconvolutional layer for
+    :class:`Conv2DLayer`, :class:`Conv1DLayer`; or
+    an unpooling layer for :class:`MaxPool2DLayer`.
+
+    It is specially useful for building (convolutional)
+    autoencoders with tied parameters.
+
+    Note that if the layer to be inverted contains a nonlinearity
+    and/or a bias, the :class:`InverseLayer` will include the derivative
+    of that in its computation.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    layer : a :class:`Layer` instance or a tuple
+        The layer with respect to which the instance of the
+        :class:`InverseLayer` is inverse to.
+
+    Examples
+    --------
+    >>> import lasagne
+    >>> from lasagne.layers import InputLayer, Conv2DLayer, DenseLayer
+    >>> from lasagne.layers import InverseLayer
+    >>> l_in = InputLayer((100, 3, 28, 28))
+    >>> l1 = Conv2DLayer(l_in, num_filters=16, filter_size=5)
+    >>> l2 = DenseLayer(l1, num_units=20)
+    >>> l_u = InverseLayer(l2, l1)  # As Deconv2DLayer
+    """
+    def __init__(self, incoming, layer, **kwargs):
+
+        super(InverseLayer, self).__init__(
+            [incoming, layer, layer.input_layer], **kwargs)
+
+    def get_output_shape_for(self, input_shapes):
+        return input_shapes[2]
+
+    def get_output_for(self, inputs, **kwargs):
+        input, layer_out, layer_in = inputs
+        return theano.grad(None, wrt=layer_in, known_grads={layer_out: input})
+
+
+def build_autoencoder(layer, nonlinearity='same', b='new'):
+    """
+    Given a :class:`Layer` instance, this function builds a
+    symmetric autoencoder with tied weights.
```

Comment:
Again, it would be good to have a very short sentence so it fits the overview tables. Maybe: "Unfolds a stack of layers into a symmetric autoencoder with tied weights." (Okay, that's not much shorter, but it gives some more information in the same length.)